untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,552 @@
+
+## A bit of history
+
+> 1936 - Alan Turing invents every programming language that will ever be but is shanghaied by British Intelligence to be 007 before he can patent them.
+>
+> 1936 - Alonzo Church also invents every language that will ever be but does it better.
+> His lambda calculus is ignored because it is insufficiently C-like.
+> This criticism occurs in spite of the fact that C has not yet been invented.
+>
+> -- James Iry, [A Brief, Incomplete, and Mostly Wrong History of Programming Languages][brief-incomplete-history]
+
+[brief-incomplete-history]: http://james-iry.blogspot.com/2009/05/brief-incomplete-and-mostly-wrong.html
+
+Lambda calculus, initially envisioned as a formal logic system, was developed by [Alonzo Church](https://en.wikipedia.org/wiki/Alonzo_Church) around the 1930s to explore the foundations of mathematics.
+
+The initial formulation had a logical inconsistency known as the [Kleene–Rosser paradox].
+To sidestep this issue, Church isolated the part of lambda calculus relevant only to computation in 1936.
+This isolate is now known as the untyped lambda calculus.
+
+Later, in 1940, a typed version of lambda calculus based on Russel's type theory was introduced.
+It has weaker expressive power, but it's logically consistent.
+
+[Kleene–Rosser paradox]: https://en.wikipedia.org/wiki/Kleene%E2%80%93Rosser_paradox
+
+In the mid-1960s, Peter Landin showed that lambda calculus models arbitrarily complex programming languages.
+Arguably, this insight kickstarted the research on functional programming languages.
+
+## The pure untyped lambda calculus
+
+The simplest, smallest type of lambda calculus is the pure untyped lambda calculus, so we'll start with it.
+
+Like any untyped lambda calculus, it is first and foremost a consistent system for rewriting expressions.
+And it's called pure because it doesn't have anything beyond what's strictly necessary – it only contains functions and variables.
+
+We can formally define the pure untyped lambda calculus in terms of _abstractions_ and _applications_.
+
+### Abstractions
+
+An abstraction, or a functional abstraction, is a parametric expression, that is to say, a function.
+Lambda calculus only defines univariate (single-variable) functions, but we can easily extend it to include multivariate functions too, as we'll see later.
+
+We introduce new abstractions using the $\lambda$ symbol.
+An abstraction consists of a _head_ and a _body_, separated by the dot ($.$) symbol.
+The head contains the $\lambda$ symbol and the parameter name.
+The body is an arbitrary expression.
+For example, $\lambda x. x$ is an abstraction.
+
+Lambda calculus derives its name from the $\lambda$ symbol used in this notation.
+The symbol was most likely chosen arbitrarily.
+However, there's an apocryphal story that the notation was derived from $\hat x$ morphing first into $\Lambda x$ and later into $\lambda x$.
```

Comment:
That's not quite arbitrarily. $\hat x$ has been chosen since Church was following Russell's notation from Principia Mathematica, but the editors had no $\hat$ when they were preparing Church's manuscript for publication. You may google that for a good reference.

Reply:
The reason I'm being wishy-washy here is that I wasn't able to find a good reference for this when I searched a few years back (better than nth-hand rumours and speculation). If you have one, please share. Otherwise, I'm sticking to this version.

Reply:
Perhaps we can just cut the "the symbol was most likely chosen arbitrarily. However,"?

Reply:
I've reworded the section slightly to align more with "History of λ-calculus" and added a reference in 87137c74d9b2ea5075b6f7d4169037f2b6df4472